### PR TITLE
Load contribution page if live payment processor is disabled but test is available

### PR DESCRIPTION
**Pull request by mattwire**
Overview
----------------------------------------
Contribution pages only save the ID of the live payment processor. When loading the contribution page a function checks and loads payment processors - it maps and loads test processor IDs for each live processor if available.
However, if you disable the live processor in `civicrm_payment_processor` table (set `is_active` = 0) the contribution page will not load if you specify the parameter `&action=preview` as part of the URL. This means you must keep the live processor enabled on a development site and you have to remove/change the credentials so it doesn't talk to the live payment systems.

This change removes the requirement to have the live processor enabled for a contribution page.

Before
----------------------------------------
Live processor cannot be disabled for a contribution page.

After
----------------------------------------
Live processor can be disabled for a contribution page and you can run the contribution page in test mode only.

Technical Details
----------------------------------------
`getAllPaymentProcessors` filters on is_active. We switch to retrieving directly via API the list of payment processors.

Comments
----------------------------------------
Note that this can only be triggered by modifying a payment processor via the API or directly in the database currently so should be of minimal risk

### CODE REVIEW

File: CRM/Financial/BAO/PaymentProcessor.php [hunk 1]
```diff
@@ -379,22 +379,23 @@ public static function getAllPaymentProcessors($mode = 'all', $reset = FALSE, $i
...
-        foreach ($possibleLiveIDs as $possibleLiveID) {
-          if (isset($processors[$possibleLiveID]) && ($liveProcessorName = $processors[$possibleLiveID]['name']) != FALSE) {
-            foreach ($testProcessors as $index => $testProcessor) {
-              if ($testProcessor['name'] == $liveProcessorName) {
-                $ids[] = $testProcessor['id'];
-              }
+    if (is_array($ids) && in_array('TestMode', $capabilities)) {
+      $testProcessors = self::getAllPaymentProcessors('test');
+      $allPaymentProcessors = civicrm_api3('PaymentProcessor', 'get', [
+        'return' => ["id", "name"],
```
Comment (eileenmcnaughton) [inline CRM/Financial/BAO/PaymentProcessor.php]: We should use single quotes here

**Comment by eileenmcnaughton** [comment]
@mattwire I'm not a fan of endlessly adding parameters to functions - but given getAllPaymentProcessors is cached I could make a case for adding $activeOnly or $isActive = TRUE (with maybe TRUE, FALSE or NULL supported) - & calling that rather than an api call here

**Comment by monishdeb** [comment]
@mattwire @eileenmcnaughton shouldn't we prevent disabling a payment processor in the first place if it used in online contribution page or event? Like we are doing in case of deleting/disabling price field, disabling financial type. 

If not I think the event should also be taken care of.

**Comment by eileenmcnaughton** [comment]
@monishdeb this is not really a UI config @mattwire is trying to support - but a developer / site tester config

**Comment by monishdeb** [comment]
ah ok

**Comment by mattwire** [comment]
> I'm not a fan of endlessly adding parameters to functions

@eileenmcnaughton That's how I thought about doing it originally :smile: so have now updated the PR to match

**Comment by mattwire** [comment]
test this please

**Comment by mattwire** [comment]
test this please

**Comment by jaapjansma** [comment]
Betty and I are reviewing PRs and we came across this one. We looked up the payment processor in dmaster but we could not disable the live payment processor but enable the test. The settings are both in screen. 

See screenshot below:

Screenshot_20200525_105736

**Comment by mattwire** [comment]
@jaapjansma This can only be done via the API (Paymentprocessor.create) and you specify the ID of the live processor. It is intended for use on staging/dev sites when you don't want to accidently trigger actions (eg. cancel subscription) on the live because you have copied data over from live.

**Comment by jaapjansma** [comment]
Thanks for your answer. So my assumption is then that this functionality is never used as site administrators would not know about this and would leave the live payment processor enabled in their test sites.

**Comment by mattwire** [comment]
@jaapjansma Currently it is *not* possible to disable the live processor without the test because it causes CiviCRM to crash. If this change is merged we could do it by the API, we could add it to the UI in CiviCRM core or we could write an extension that allows you to change it.

**Comment by jaapjansma** [comment]
@mattwire I have disabled the live through the API explorer. But now the live and the test are both disabled. Can you fix this?

**Comment by mattwire** [comment]
@eileenmcnaughton Can I get your thoughts on this please? As @jaapjansma points out it's not possible to disable live using the UI or the API. You *can* disable test independently using the API but the BAO toggles test `is_active` if you toggle live `is_active`. For my use-case that's ok because this was intended to allow a staging/dev-site setup where the live is disabled (and toggling the DB value directly via an extension / future setting / something seems like it may be ok).

I envisaged this PR being the "enabler" for this functionality rather than the full "selectable via UI" and that UI/API would follow later. I could write a test for what this PR contains so we have the functionality tested even if not available *yet* via the UI.

File: CRM/Financial/BAO/PaymentProcessor.php [hunk 2]
```diff
@@ -293,10 +295,13 @@ public static function getAllPaymentProcessors($mode = 'all', $reset = FALSE, $i
     }
 
     $retrievalParameters = [
-      'is_active' => TRUE,
       'options' => ['sort' => 'is_default DESC, name', 'limit' => 0],
       'api.payment_processor_type.getsingle' => 1,
     ];
+    if (isset($isActive)) {
+      // We use isset because we don't want to set the is_active parameter at all is $isActive is NULL
```
Comment (eileenmcnaughton) [inline CRM/Financial/BAO/PaymentProcessor.php:302]: if not is

**Comment by eileenmcnaughton** [comment]
@mattwire I'm OK with this change in that I think the underlying logic of the functions makes sense with it. I don't think we should go down the path of exposing in the UI or even the API - I feel like that will just grow & grow & get nasty. But as a 'hey if you have sql access you can do this not-particularly-supported-thing'  I'm Ok with that as the function change is logical.

However, in tests I found that there is a bug with this in that it no longer returns processors from other domains when specified by id - I will something up with my test / the cut down fix

**Comment by eileenmcnaughton** [comment]
Closing in favour of https://github.com/civicrm/civicrm-core/pull/17828 which doesn't have the regression

**Pull request closed without merging**